autopaho/backoff.go

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v2.0
+ *  and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    https://www.eclipse.org/legal/epl-2.0/
+ *  and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ *  SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+package autopaho
+
+import (
+	"math/rand"
+	"time"
+)
+
+// The BackoffStrategy is a container for the configuration of a given strategy whereas the returned Backoff is the actual implementation
+// As such BackoffStrategy instances are supposed to be thread safe and freely sharable between users.
+type BackoffStrategy interface {
+	// Returns a configured backoff instance.
+	GetBackoff() Backoff
+}
+
+// Backoff represents a configured instance of the strategy which computes the next backoff duration.
+// Instances are NOT supposed to be shared.
+type Backoff interface {
+	// Returns the next backoff duration.
+	Next() time.Duration
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// implementation for constant backoff
+////////////////////////////////////////////////////////////////////////////////
+
+// The ConstantBackoff implements the Backoff interface and provides a constant duration.
+type ConstantBackoff struct {
+	Backoff
+	_Duration time.Duration
+}
+
+// Creates a new ConstantBackoff with the specified duration used as the constant backoff.
+func NewConstantBackoff(duration time.Duration) *ConstantBackoff {
+	return &ConstantBackoff{
+		_Duration: duration,
+	}
+}
+
+func (c *ConstantBackoff) Next() time.Duration {
+	return c._Duration
+}
+
+// The ConstantBackoffStrategy implements the BackoffStrategy interface and provides instances a constant duration.
+type ConstantBackoffStrategy struct {
+	BackoffStrategy
```

Comment:
Why embed `BackoffStrategy`? I suspect this may be a fallback to Java where you would need to declare what interface you are implementing but it's not needed here (and is harmful because it would prevent the compiler from warning you if your implementation of `GetBackoff()` has the wrong signature.

Reply:
Many thanks for pointintg this out. I wasn't aware this would be actuall harmful. This was definitely my Java way of thinking...